design/src/rfcs/rfc_0020_endpoints_20.md

```diff
@@ -0,0 +1,597 @@
+                rust("// ${section.request}.properties_mut().insert(params);")
+            }
+            else -> emptySection
+        }
+    }
+
+    private fun builderFields(section: OperationSection.MutateInput) = writable {
+        val memberParams = idx.getContextParams(operationShape)
+        val builtInParams = ruleset.parameters.toList().filter { it.isBuiltIn }
+        // first load builtins and their defaults
+        builtInParams.forEach { param ->
+            val defaultProviders = section.endpointCustomizations.mapNotNull { it.defaultFor(param, section.config) }
+            if (defaultProviders.size > 1) {
+                error("Multiple providers provided a value for the builtin $param")
+            }
+            defaultProviders.firstOrNull()?.also { defaultValue ->
+                rust(".set_${param.name.rustName()}(#W)", defaultValue)
+            }
+        }
+        // these can be overridden with client context params
+        idx.getClientContextParams(ctx.serviceShape).forEach { (name, _param) ->
+            rust(".set_${name.toSnakeCase()}(${section.config}.${name.toSnakeCase()}.as_ref())")
+        }
+
+        // lastly, allow these to be overridden by members
+        memberParams.forEach { (memberShape, param) ->
+            rust(".set_${param.name.toSnakeCase()}(${section.input}.${ctx.symbolProvider.toMemberName(memberShape)}.as_ref())")
+        }
+        rust(".build()")
+    }
+}
+```
+
+</details>
+
+#### Loading values for builtIns
+
+The fundamental point of builtIn values is enabling _other codegenerators to define where these values come from.
+Because of that,
+we will need to expose the ability to customize builtIns. One way to do this is with a new customization
+type, `EndpointCustomization`:
+
+```kotlin
+fun endpointCustomizations(
+    clientCodegenContext: C,
+    operation: OperationShape,
+    baseCustomizations: List<EndpointCustomization>
+): List<EndpointCustomization> = baseCustomizations
+
+
+abstract class EndpointCustomization {
+    abstract fun defaultFor(parameter: Parameter, config: String): Writable?
+}
+```
+
+Customizations have the ability to specify the default value for a parameter. (Of course, these customizations need to
+be wired in properly.)
+
+### Converting a Smithy Endpoint to an AWS Endpoint
+
+A Smithy endpoint has an untyped, string->`Document` collection of properties. We need to interpret these properties to
+handle actually resolving an endpoint. We will implement this by implementing `TryFrom` from the Smithy endpoint to the
+AWS endpoint:
+> Note: It's possible `AWSEndpoint` will be removed entirely and the concept will be solely expressed in middleware. If
+> this is the case, this code would be ported into middleware directly.
+
+<details>
+<summary>Converting from smithy-native `Endpoint` to `AwsEndpoint`</summary>
+
+```rust
+impl TryFrom<Endpoint> for AwsEndpoint {
+    type Error = Box<dyn std::error::Error + Send + Sync>;
```

Comment:
Should we also add a static bound here?

Reply:
static is implied in this case since it's in a box